BrainPortal/cbrain_plugins/cbrain-plugins-base/userfiles/cbrain_file_list/cbrain_file_list.rb

```diff
@@ -0,0 +1,146 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+# Model for a list of CBRAIN files as a plain text CSV file
+class CbrainFileList < CSVFile
+
+  Revision_info=CbrainFileRevision[__FILE__] #:nodoc:
+
+  # In this model we use the comma as separator, double-quote as quoting character.
+  FIELD_SEPARATOR   = ','
+  QUOTING_CHARACTER = '"'
+  RECORD_TERMINATOR = "\n"
+
+  # Structure of the CSV file; only the ID is used when this object is used as input to something else.
+  # When displayed in a web page, the associations to other models are shown by name.
+  ATTRIBUTES_LIST   = [ :id, :name, :size, :type, :data_provider_id, :user_id, :group_id ]
+
+  has_viewer :name => 'CBRAIN File List', :partial  => :cb_file_list, :if => :is_viewable?
+
+  def self.pretty_type #:nodoc:
+    "CBRAIN List of files"
+  end
+
+  #def self.file_name_pattern #:nodoc:
+  #  /\.csv$/i
+  #end
+
+  # Returns an array of arrays with the parsed content of the CSV file.
+  # No validation is done outside of having the proper general format.
+  # Entries can contain empty strings or nils.
+  #
+  #  [ [ 12, "filename1.txt", 2324, "TextFile"  ],
+  #    [ "", "",               222, ""          ],
+  #    [ 45, "image.jpg",    13121, "ImageFile" ]  ]
+  def cached_csv_array
+    @rows ||= create_csv_array(QUOTING_CHARACTER, FIELD_SEPARATOR)
+  end
+
+  # Returns an array of the IDs of the first column of the CSV file
+  # as extracted by cached_csv_array(). IDs will be numeric, or for
+  # missing rows, will contain nils. IDs can be zero.
+  #
+  #  [ 12, 0, 45, nil nil, 433 ]
+  def ordered_raw_ids
+    ids_with_zeros_and_nils = cached_csv_array.map do |row|
+      myid = row[0] # can be nil
+      if (myid.present?) && (myid =~ /^\s*\d+\s*$/)
+        myid.strip.to_i
+      else
+        nil
+      end
+    end
+    ids_with_zeros_and_nils  # [ 1234, 0, nil, 533, 243, nil, nil ] etc
+  end
+
+  # Returns the userfiles accessible to +user+, fetched by ID in the
+  # same order as from ordered_raw_ids(). Entries with a numeric ID of 0
+  # will be replaced by +missing+, (defaults to nil) and numeric entries pointing
+  # to files that are inaccessible to the user will be replaced by +invalid+
+  # (also defaults to nil). This method will raise an exception if
+  # the content of the CSV file is invalid.
+  def userfiles_accessibles_by!(user, missing=nil, invalid=nil)
```

Comment:
I will rename the methods `userfiles_accessible_by_user()`, so that it corresponds a little bit more to the other method names we have in our models...